llvm/lib/Target/RISCV/RISCVAsmPrinter.cpp

```diff
@@ -385,8 +385,32 @@ void RISCVAsmPrinter::emitStartOfAsmFile(Module &M) {
   if (const MDString *ModuleTargetABI =
           dyn_cast_or_null<MDString>(M.getModuleFlag("target-abi")))
     RTS.setTargetABI(RISCVABI::getTargetABI(ModuleTargetABI->getString()));
+
+  MCSubtargetInfo SubtargetInfo = *TM.getMCSubtargetInfo();
+
+  // Use module flag to update feature bits.
+  if (auto *MD = dyn_cast_or_null<MDNode>(M.getModuleFlag("riscv-isa"))) {
+    for (auto &ISA : MD->operands()) {
+      if (auto *ISAString = dyn_cast_or_null<MDString>(ISA)) {
+        auto ParseResult = llvm::RISCVISAInfo::parseArchString(
+            ISAString->getString(), /*EnableExperimentalExtension=*/true,
+            /*ExperimentalExtensionVersionCheck=*/true);
+        if (!errorToBool(ParseResult.takeError())) {
+          auto &ISAInfo = *ParseResult;
+          for (const auto &Feature : RISCVFeatureKV) {
+            if (ISAInfo->hasExtension(Feature.Key) &&
```

Comment:
Is it possible to report errors when two `riscv-isa` are not compatible? Like `Zdinx` and `d`, etc.

Reply:
Not without building a new ISA string and reparsing it. Do we error for that in non-LTO?

Reply:
Yes, I think we will check that (in `RISCVISAInfo::checkDependency()`).
```shell
~/workspace# clang -march=rv64imafdc_zdinx -### -
clang: error: -E or -x required when input is from standard input
clang: error: invalid arch name 'rv64imafdc_zdinx', 'f' and 'zfinx' extensions are incompatible
```

Reply:
> building a new ISA string and reparsing it

Reply:
> Yes, I think we will check that (in `RISCVISAInfo::checkDependency()`).
> 
> ```shell
> ~/workspace# clang -march=rv64imafdc_zdinx -### -
> clang: error: -E or -x required when input is from standard input
> clang: error: invalid arch name 'rv64imafdc_zdinx', 'f' and 'zfinx' extensions are incompatible
> ```

I meant during linking. That's what this code is emulating, linking two different modules in LTO.

Reply:
It's been a somewhat contentious issue. I *believe* it's still an error.

Reply:
Is this blocking for this patch? We haven't had any way of detecting such an error in LTO for a long time now.

Reply:
Maybe we should put it on the sync-up agenda. :-)

Reply:
My personal view is that if we can stop the user from doing something that's _definitely_ nonsensical then it's nice to do so, but it needn't be blocking unless we think it's likely to be a common error. We have had issues in this general area of error checking being overzealous leading to compatibility issues, so I'd rather we fail to catch some silly things vs incorrectly erroring on reasonable things and leaving users stuck until a toolchain update.

Reply:
The consensus from the sync-up call was I think that:
* This patch shouldn't be blocked on adding additional error checking for incompatible combinations. We'd previously identified warnings for this to be nice to have, but non-essential.
* There's quite a lot of subtlety about what errors and the potential impact it has - in fact it seems we currently error when combining objects with D and Zdinx and arguably shouldn't. This probably deserves more testing and discussion, but separately to this patch.